### PR TITLE
Training Multi-Class Logistic Regression model with only shared features. 

### PR DESCRIPTION
**Pull request by axiomcura**
In the previous pull request (#8), a significant number of features were found to not overlap between the JUMP and Cell-injury image-based profiles (#9). This discrepancy is likely caused by the usage of different CellProfiler versions in both projects. The Cell-injury project, the paper states that CellProfiler 2.1.2 was used for extracting morphological features, whereas the JUMP dataset used v4.1.2.

This pull request introduces new changes in the feature selection module to facilitate training our models only with shared features. Previously, the workflow followed the order of feature selection, model training/testing, model evaluation, finding shared features, and then application to the JUMP dataset. However, the updated workflow now begins with finding shared features, followed by feature selection, model training/testing, model evaluation, and finally application to the JUMP dataset.

With the shift to using only shared features prior to feature selection, the parameters for feature selection were significantly affected. The original parameters resulted in the removal of all features. Therefore, I used  `Pycytominer`'s default parameters for feature selection, which retained of most morphology features.

Furthermore, the figures have been updated in this pull request to reflect these changes. Unfortunately, the confusion matrix depicting our models performance predicting the overlapping treatments found in JUMP performed poorly. 

**Other changes**
- bug fixes in the `utils.py` module 
- formatting
- More documentation added

### CODE REVIEW

File: data/JUMP_data/nbconverted/download.py [hunk 1]
```diff
@@ -1,54 +1,76 @@
...
-            for chunk in response.iter_content(chunk_size=8192):
-                f.write(chunk)
+# save concatenated df into ./data/JUMP_data folders
+jump_df.to_csv(
+    "JUMP_all_plates_normalized_negcon.csv.gz", index=False, compression="gzip"
+)
 
 
-# In[ ]:
+# Here, we obtain information about the feature space by splitting both the meta and feature column names and storing them in a dictionary. This dictionary holds information about the feature space and will be utilized for downstream analysis when identifying shared features across different datasets, such as the Cell-injury dataset.
```
Comment (MikeLippincott) [inline data/JUMP_data/nbconverted/download.py]:
```suggestion
# Here, we obtain information about the feature space by splitting both the meta and feature column names and storing them in a dictionary. 
This dictionary holds information about the feature space and will be utilized for downstream analysis when identifying shared features across different datasets, such as the Cell-injury dataset.
```

File: notebooks/1.data_splits/nbconverted/1.data_splits.py [hunk 2]
```diff
@@ -175,53 +174,10 @@
...
-with open(data_split_dir / "plate_info.json", mode="w") as stream:
+with open(data_split_dir / "cell_injury_plate_info.json", mode="w") as stream:
     json.dump(plate_meta, stream)
 
 
-# Set numerical labels for the treatment
-
-# In[9]:
-
-
```
Comment (MikeLippincott) [inline notebooks/1.data_splits/nbconverted/1.data_splits.py:186]: Why removed labeler here?

File: notebooks/2.modeling/nbconverted/2.modeling.py [hunk 3]
```diff
@@ -1,6 +1,10 @@
 #!/usr/bin/env python
 # coding: utf-8
 
+# # Module 2: Modeling
+#
+# In this notebook, we focus on developing and training a Multi-class logistic regression model, employing Randomized Cross-Validation (CV) for hyperparameter tuning to address our classification task. The dataset is split into an 80% training set and a 20% testing set. To evaluate the performance of our model during training, we used performance evaluation metrics such as precision, recall, and F1 scores. Additionally, we extend our evaluation by testing our model on a holdout dataset, which includes plate, treatment, and well information, providing a comprehensive assessment of its real-world performance.
```
Comment (MikeLippincott) [inline notebooks/2.modeling/nbconverted/2.modeling.py:6]:
```suggestion
# In this notebook, we focus on developing and training a Multi-class logistic regression model, employing Randomized Cross-Validation (CV) for hyperparameter tuning to address our classification task. The dataset is split into an 80% training set and a 20% testing set. 
To evaluate the performance of our model during training, we used performance evaluation metrics such as precision, recall, and F1 scores. 
Additionally, we extend our evaluation by testing our model on a holdout dataset, which includes plate, treatment, and well information, providing a comprehensive assessment of its real-world performance.
```

File: notebooks/2.modeling/nbconverted/2.modeling.py [hunk 4]
```diff
@@ -78,22 +87,33 @@
...
 
+
 # spliting meta and feature column names
 _, feat_cols = split_meta_and_features(X_train)
 
+# checking if the feature space are identical (also looks for feature space order)
+assert check_feature_order(
+    ref_feat_order=shared_features, input_feat_order=X_test.columns.tolist()
+), "Feature space are not identical"
+
```
Comment (MikeLippincott) [inline notebooks/2.modeling/nbconverted/2.modeling.py:116]: The assertion will not pass if the order changes???

File: notebooks/4.visualization/confusion_matrix/confusion_matrix.png
Comment (MikeLippincott) [inline notebooks/4.visualization/confusion_matrix/confusion_matrix.png:1]: The model did not seem to change drastically

File: notebooks/4.visualization/confusion_matrix/overlapping_confusion_matrix.png
Comment (MikeLippincott) [inline notebooks/4.visualization/confusion_matrix/overlapping_confusion_matrix.png:1]: Is the model not generalizing to JUMP or could there be an error and the Not Shuffled model is actually the shuffled model? They look the exact same in terms of performance.

Also, every class is represented in the JUMP dataset? Did we talk about this ever? Were some classes missing???

If this is the case the row total for the missing classes should be 0.

**Review by MikeLippincott — APPROVED** [review]
LGTM. The plots look a bit funky. Consider double checking the JUMP testing set plots. Other than that, ready to merge!

File: notebooks/1.data_splits/nbconverted/1.data_splits.py (hunk 2 above)
Comment (axiomcura) [reply notebooks/1.data_splits/nbconverted/1.data_splits.py:186]: I moved it to the `0.feature_selection` directory for convenience. I aimed to capture the feature space right from the start, believing it to be an easier approach.

File: notebooks/4.visualization/confusion_matrix/confusion_matrix.png
Comment (axiomcura) [reply notebooks/4.visualization/confusion_matrix/confusion_matrix.png:1]: Correct, when reducing the dataset to only the shared features, the model continued to perform well. However, a the issue with the treatment holdout still exists. The challenge here is that only a limited number of treatments were withheld due to the heuristic we applied for selecting treatment holdouts (explained in the data split notebook). Unfortunately, not all treatments known to cause an injury were included in the selection.

File: notebooks/4.visualization/confusion_matrix/overlapping_confusion_matrix.png
Comment (axiomcura) [reply notebooks/4.visualization/confusion_matrix/overlapping_confusion_matrix.png:1]: In the `3.jump_analysis.ipynb` module, the `InChIKey` was used  to determine the treatments present in the JUMP dataset. The `InChIKey` serves as a unique identifier for individual compounds. Hence, we utilized it from both the cell-injury and JUMP datasets, cross-referencing them to identify shared `InChIKeys`. Below is an image illustrating the treatments present in the JUMP dataset along with the corresponding number of wells.

image

All in all, we identified only four treatments, each inducing different injuries. Despite DMSO having the largest number of wells than the rest, this shouldn't affect our analysis because our model was trained using the  `class_weight=balanced`.

I'll give it another look to see if there's a small bug in the analysis

File: notebooks/2.modeling/nbconverted/2.modeling.py (hunk 4 above)
Comment (axiomcura) [reply notebooks/2.modeling/nbconverted/2.modeling.py:116]: Correct. Order of the feature space does matter in this case. 

When a machine learning model is trained, it learns patterns based on the order of the features in the training data. If the order of the features in the JUMP dataset does not match the order of the features in the training data, the model may not perform as expected.

**Comment by axiomcura** [comment]
@MikeLippincott Thank you for your comments and suggestions. I will be merging this PR. Please feel free to continue our discussion if needed!
